modules/raft/src/main/java/org/apache/ignite/raft/jraft/entity/codec/v1/V1Encoder.java

```diff
@@ -46,130 +48,137 @@ public byte[] encode(final LogEntry log) {
-            final byte[] ps = AsciiStringUtil.unsafeEncode(peerStr);
-            Bits.putShort(content, pos, (short) peerStr.length());
-            System.arraycopy(ps, 0, content, pos + 2, ps.length);
-            pos += 2 + ps.length;
+
+        return content;
+    }
+
+    private static int nodesListSizeInBytes(@Nullable List<PeerId> nodes, List<String> nodeStrs) {
+        int size = 0;
+
+        if (nodes != null) {
+            for (PeerId node : nodes) {
+                String nodeStr = node.toString();
+
+                nodeStrs.add(nodeStr);
+                size += 2 + nodeStr.length();
+            }
         }
-        // old learners
-        // old learners count
-        Bits.putInt(content, pos, oldLearnerCount);
-        pos += 4;
-        for (final String peerStr : oldLearnerStrs) {
-            final byte[] ps = AsciiStringUtil.unsafeEncode(peerStr);
-            Bits.putShort(content, pos, (short) peerStr.length());
-            System.arraycopy(ps, 0, content, pos + 2, ps.length);
-            pos += 2 + ps.length;
+
+        return size + sizeInBytes(nodeStrs.size());
+    }
+
+    private static int writeNodesList(int pos, byte[] content, List<String> nodeStrs) {
+        pos = writeLong(nodeStrs.size(), content, pos);
+
+        for (String nodeStr : nodeStrs) {
+            int length = nodeStr.length();
+
+            Bits.putShort(content, pos, (short) length);
+            pos += 2;
+
+            AsciiStringUtil.unsafeEncode(nodeStr, content, pos);
+            pos += length;
         }
-        // data
-        if (data != null) {
-            System.arraycopy(data.array(), data.position(), content, pos, data.remaining());
+
+        return pos;
+    }
+
+    // Based on DirectByteBufferStreamImplV1.
+    private static int writeLong(long val, byte[] out, int pos) {
+        while ((val & 0xFFFF_FFFF_FFFF_FF80L) != 0) {
+            byte b = (byte) (val | 0x80);
+
+            out[pos++] = b;
+
+            val >>>= 7;
         }
 
-        return content;
+        out[pos++] = (byte) val;
+
+        return pos;
+    }
+
+    /**
+     * Returns the number of bytes, required by the {@link #writeLong(long, byte[], int)} to write the value.
+    */
+    private static int sizeInBytes(long val) {
+        if (val >= 0) {
+            if (val < (1L << 7)) {
+                return 1;
+            } else if (val < (1L << 14)) {
+                return 2;
+            } else if (val < (1L << 21)) {
+                return 3;
+            } else if (val < (1L << 28)) {
+                return 4;
+            } else if (val < (1L << 35)) {
+                return 5;
+            } else if (val < (1L << 42)) {
+                return 6;
+            } else if (val < (1L << 49)) {
+                return 7;
+            } else if (val < (1L << 56)) {
+                return 8;
+            } else {
+                return 9;
+            }
+        } else {
+            return 10;
+        }
     }
```

Comment:
```suggestion
    private static int sizeInBytes(long val) {
        if (val >= 0) {
            int leadingZeros = Long.numberOfLeadingZeros(val);
            int bits = 64 - leadingZeros;
            return (bits + 6) / 7;
        } else {
            return 10;
        }
    }
```

Reply:
It's going to be slower on smaller values